src/network/networkd-link.c

```diff
@@ -2346,163 +2296,271 @@ static int link_update_flags(Link *link, sd_netlink_message *m, bool force_updat
 
-                r = ipv4ll_update_mac(link);
-                if (r < 0)
-                        return log_link_warning_errno(link, r, "Could not update MAC address in IPv4LL client: %m");
+        link_drop_from_master(link);
 
-                r = dhcp4_update_mac(link);
-                if (r < 0)
-                        return log_link_warning_errno(link, r, "Could not update MAC address in DHCP client: %m");
+        link->master_ifindex = master_ifindex;
 
-                r = dhcp6_update_mac(link);
-                if (r < 0)
-                        return log_link_warning_errno(link, r, "Could not update MAC address in DHCPv6 client: %m");
+        r = link_append_to_master(link);
+        if (r < 0)
+                return log_link_debug_errno(link, r, "Failed to append link to master: %m");
 
-                r = radv_update_mac(link);
-                if (r < 0)
-                        return log_link_warning_errno(link, r, "Could not update MAC address for Router Advertisement: %m");
+        return 0;
+}
 
-                if (link->ndisc) {
-                        r = sd_ndisc_set_mac(link->ndisc, &link->hw_addr.addr.ether);
-                        if (r < 0)
-                                return log_link_warning_errno(link, r, "Could not update MAC for NDisc: %m");
-                }
+static int link_update_hardware_address(Link *link, sd_netlink_message *message) {
+        hw_addr_data hw_addr;
+        int r;
 
-                if (link->lldp) {
-                        r = sd_lldp_set_filter_address(link->lldp, &link->hw_addr.addr.ether);
-                        if (r < 0)
-                                return log_link_warning_errno(link, r, "Could not update MAC address for LLDP: %m");
-                }
+        assert(link);
+        assert(message);
 
-                r = ipv4_dad_update_mac(link);
-                if (r < 0)
-                        return log_link_warning_errno(link, r, "Could not update MAC address in IPv4 ACD client: %m");
-        }
+        r = netlink_message_read_hw_addr(message, IFLA_BROADCAST, &link->bcast_addr);
+        if (r < 0 && r != -ENODATA)
+                return log_link_debug_errno(link, r, "rtnl: failed to read broadcast address: %m");
 
-        old_master = link->master_ifindex;
-        (void) sd_netlink_message_read_u32(m, IFLA_MASTER, (uint32_t *) &link->master_ifindex);
+        r = netlink_message_read_hw_addr(message, IFLA_ADDRESS, &hw_addr);
+        if (r == -ENODATA)
+                return 0;
+        if (r < 0)
+                return log_link_warning_errno(link, r, "rtnl: failed to read hardware address: %m");
 
-        link_was_admin_up = link->flags & IFF_UP;
-        had_carrier = link_has_carrier(link);
+        if (link->hw_addr.length == hw_addr.length &&
+            memcmp(link->hw_addr.addr.bytes, hw_addr.addr.bytes, hw_addr.length) == 0)
```

Comment:
we should have a helper hw_addr_data_equal()...

(but i understand this wasn#t inroduced in this commit)